### PR TITLE
Init

### PR DESCRIPTION
**Pull request by Scriptkiddi**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
+# CertificatesRole Name
+
+This role install a Certificate Authoriy and generates certificates for every host in the inventory and copys them over
```
Comment (dasJ) [inline README.md]: *installs

File: README.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+## Role Variables
+
+### General
+
+| Name                              | Required/Default              | Description                                                    |
+|-----------------------------------|:-----------------------------:|----------------------------------------------------------------|
+| certificates_ca_server            | :heavy_check_mark:            | Inventory Hostname of the server running the ca                |
+| certificates_ca_key_dir           | `/etc/ssl/private/ansible-ca` | Directory where the files for the ca should be stored          |
+| certificates_local_ca_path        | `/etc/ssl/ansible-ca.crt`     | Path where the ca.crt should be copied to on the localhost     |
```
Comment (dasJ) [inline README.md]: What is localhost?

File: README.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+| certificates_ca_common_name       | :heavy_check_mark:            | Common Name for the CA                                         |
+| certificates_ca_country_name      | :heavy_check_mark:            | Country where the CA ist located                               |
+| certificates_ca_email_address     | :heavy_check_mark:            | Email Address where the CA Administrator can be reached        |
+| certificates_local_key_dir        | `/etc/ssl/private/ansible`    | Directory where the files for the localhost should be stored   |
+| certificates_common_name          | `{{ inventory_hostname }}`    | Common Name for the certificate for the localhost              |
+| certificates_country_name         | :heavy_check_mark:            | Country where the Host is located                              |
+| certificates_email_address        | :heavy_check_mark:            | Email Address where the localhost Administrator can be reached |
+| certificates_organization_name    | :heavy_check_mark:            | Organization Name that runs the localhost                      |
```
Comment (dasJ) [inline README.md]: Is there any way I can set key the length?

File: meta/main.yml [hunk 4]
```diff
@@ -0,0 +1,20 @@
+---
+galaxy_info:
+  author: Insert author(s) here
```
Comment (dasJ) [inline meta/main.yml]: ...

File: tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,65 @@
+---
+- name: Install pythonOpenSSl
+  apt:
+    name: python-openssl
```
Comment (dasJ) [inline tasks/main.yml:4]: You missed the blank line

File: templates/ca.conf.j2 [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+[ CA_default ]
+dir = {{ certificates_ca_key_dir }}
+certs = $dir/
+new_certs_dir = $dir/
+database = $dir/index.txt
+crlnumber = $dir/crl_number
+certificate = $dir/ca.crt
+private_key = $dir/ca-key.pem
+default_days = 3650
```
Comment (dasJ) [inline templates/ca.conf.j2]: It would be nice if I could configure the days

File: templates/revoke.sh.j2 [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+crl_out () {
+  # show certificate revocation list
+  openssl crl -in ca-crl.pem -noout -text
+}
+
+crl_verify () {
+  # verify that a certificate has been revoked
+  temp_pem="${mktemp}"
```
Comment (dasJ) [inline templates/revoke.sh.j2]: Is this even tested? `${` uses a variable, `$(` executes a command

**Review by dasJ — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (Scriptkiddi) [reply README.md]: fixed

File: README.md (hunk 2 above)
Comment (Scriptkiddi) [reply README.md]: fixed

File: templates/revoke.sh.j2 (hunk 7 above)
Comment (Scriptkiddi) [reply templates/revoke.sh.j2]: yes its tested

File: README.md (hunk 3 above)
Comment (Scriptkiddi) [reply README.md]: now there is

File: tasks/main.yml (hunk 5 above)
Comment (Scriptkiddi) [reply tasks/main.yml:4]: fixed

File: templates/ca.conf.j2 (hunk 6 above)
Comment (Scriptkiddi) [reply templates/ca.conf.j2]: fixed

**Comment by Scriptkiddi** [comment]
@dasJ @schuppentier pls rereview

File: README.md [hunk 8]
```diff
@@ -0,0 +1,54 @@
+# CertificatesRole Name
```
Comment (dasJ) [inline README.md]: ?

File: meta/main.yml [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+  platforms:
+    - name: Ubuntu
+      versions:
+        - xenial
+        - bionic
+
+  galaxy_tags:
+    - CA
+    - Certificate
+    - Authority
```
Comment (dasJ) [inline meta/main.yml:20]: ... openssl?

File: tasks/generate_ca.yml [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+    owner: root
+    group: root
+    mode: 744
+  delegate_to: "{{ certificates_ca_server }}"
+
+- name: Create initial certificate revocation list squence number
+  shell: "echo 00 > crl_number"
+  args:
+    chdir: "{{certificates_ca_key_dir}}"
+    creates: crl_number
```
Comment (dasJ) [inline tasks/generate_ca.yml:45]: Are you sure this works with a relative path?

File: tasks/generate_ca.yml [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+  delegate_to: "{{ certificates_ca_server }}"
+
+- name: Install revocation script
+  template:
+    src: revoke.sh.j2
+    dest: "{{certificates_ca_key_dir}}/revoke.sh"
+    owner: root
+    group: root
+    mode: 744
```
Comment (dasJ) [inline tasks/generate_ca.yml]: Why not 755? Also you forgot the 0 at the beginning

File: tasks/get_ca_certificate.yml [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+- name: Slurp ca certificate
+  slurp:
+    src: "{{ certificates_ca_key_dir }}/ca.crt"
+  register: ca_certificate
+  delegate_to: "{{ certificates_ca_server }}"
+
+- name: Copy ca certificate to local host
+  copy:
+    content: "{{ ca_certificate['content'] | b64decode }}"
+    dest: "{{ certificates_local_ca_path }}"
```
Comment (dasJ) [inline tasks/get_ca_certificate.yml:22]: Don't you want to set some permissions on this?

File: tasks/get_crl.yml [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+- name: Slurp crl
+  slurp:
+    src: "{{ certificates_ca_key_dir }}/ca-crl.pem"
+  register: crl_certificate
+  delegate_to: "{{ certificates_ca_server }}"
+
+- name: Copy crl to local host
+  copy:
+    content: "{{ crl_certificate['content'] | b64decode }}"
+    dest: "{{ certificates_local_key_dir }}/ca-crl.pem"
```
Comment (dasJ) [inline tasks/get_crl.yml:11]: Also - no permissions at all?

File: tasks/get_crl.yml [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+- name: Slurp crl_number
+  slurp:
+    src: "{{ certificates_ca_key_dir }}/crl_number"
+  register: crl_number
+  delegate_to: "{{ certificates_ca_server }}"
+
+- name: Copy crl_number to local host
+  copy:
+    content: "{{ crl_number['content'] | b64decode }}"
+    dest: "{{ certificates_local_key_dir }}/crl_number"
```
Comment (dasJ) [inline tasks/get_crl.yml:23]: Didn't know you need that

File: tasks/get_new_certificate.yml [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+- name: Slurp certificate
+  slurp:
+    src: "{{ certificates_ca_key_dir }}/{{ inventory_hostname }}.crt"
+  register: signed_certificate
+  delegate_to: "{{ certificates_ca_server }}"
+
+- name: Copy certificate to localhost
+  copy:
+    content: "{{ signed_certificate['content'] | b64decode }}"
+    dest: "{{ certificates_local_key_dir }}/{{ inventory_hostname }}.crt"
```
Comment (dasJ) [inline tasks/get_new_certificate.yml:54]: ... permissions?

File: templates/ca.conf.j2 [hunk 16]
```diff
@@ -0,0 +1,20 @@
+
```
Comment (dasJ) [inline templates/ca.conf.j2:2]: `{{ ansible_managed }}`

**Review by dasJ — CHANGES_REQUESTED** [review]

File: README.md (hunk 8 above)
Comment (Scriptkiddi) [reply README.md]: fixed

File: meta/main.yml (hunk 9 above)
Comment (Scriptkiddi) [reply meta/main.yml:20]: done

File: tasks/generate_ca.yml (hunk 10 above)
Comment (Scriptkiddi) [reply tasks/generate_ca.yml:45]: yes since i pass chdir

File: tasks/get_crl.yml (hunk 14 above)
Comment (Scriptkiddi) [reply tasks/get_crl.yml:23]: i use it to determine if i need to update the crl

File: templates/ca.conf.j2 (hunk 16 above)
Comment (Scriptkiddi) [reply templates/ca.conf.j2:2]: fixed

File: tasks/generate_ca.yml (hunk 11 above)
Comment (Scriptkiddi) [reply tasks/generate_ca.yml]: fixed

File: tasks/get_ca_certificate.yml (hunk 12 above)
Comment (Scriptkiddi) [reply tasks/get_ca_certificate.yml:22]: fixed

File: tasks/get_new_certificate.yml (hunk 15 above)
Comment (Scriptkiddi) [reply tasks/get_new_certificate.yml:54]: fixed

**Comment by Scriptkiddi** [comment]
@schuppentier @dasJ pls rereview

**Review by dasJ — APPROVED** [review]

**Pull request merged**